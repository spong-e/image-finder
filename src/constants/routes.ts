const ROUTES = {
  ENTRY: "/",
  DISPLAY: "/display",
  SEARCH: "/search",
};

export { ROUTES };
