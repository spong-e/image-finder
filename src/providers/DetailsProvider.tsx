import * as React from "react";
import { createContext, Dispatch, ReactNode, useReducer } from "react";
import detailsReducer from "./DetailsReducer";

const DetailsContext = createContext<DetailsState | undefined>(undefined);

const DetailsDispatchContext = createContext<
  Dispatch<DetailsAction> | undefined
>(undefined);

interface DetailsProviderProps {
  children: ReactNode;
}

function DetailsProvider({ children }: DetailsProviderProps) {
  const initialState: DetailsState = {
    details: {
      firstName: "",
      surname: "",
      topic: "",
      thumbnail: "",
    },
  };

  const [state, dispatch] = useReducer(detailsReducer, initialState);

  // NOTE: you *might* need to memoize this value
  // Learn more in http://kcd.im/optimize-context

  const value = { state, dispatch };
  return (
    <DetailsContext.Provider value={state}>
      <DetailsDispatchContext.Provider value={dispatch}>
        <>
          {JSON.stringify(state, null, 2)}
          {children}
        </>
      </DetailsDispatchContext.Provider>
    </DetailsContext.Provider>
  );
}

export { DetailsProvider, DetailsContext, DetailsDispatchContext };
