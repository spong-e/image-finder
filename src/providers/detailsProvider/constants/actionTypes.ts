const ACTION_TYPES = {
  SET_DETAILS: "SET_DETAILS",
  IMAGE_SEARCH: "IMAGE_SEARCH",
};

export { ACTION_TYPES };
