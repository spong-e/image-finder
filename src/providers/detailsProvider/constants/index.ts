export { ACTION_TYPES } from "./actionTypes";
