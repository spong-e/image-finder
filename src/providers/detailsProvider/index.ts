export * from "./constants";
export * from "./context";
export * from "./hooks";
