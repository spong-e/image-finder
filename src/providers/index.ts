export { DetailsProvider } from "./DetailsProvider";
export { default as useDetails } from "./useDetails";
export { default as useDetailsActions } from "./useDetailsActions";
