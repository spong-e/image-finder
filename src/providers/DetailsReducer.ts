import produce from "immer";
import { Reducer } from "react";
import axios from "axios";
import { createApi } from "unsplash-js";
import { ACTION_TYPES } from "./constants";

const detailsReducer: Reducer<DetailsState, DetailsAction> = produce(
  (draft: DetailsState, action: DetailsAction) => {
    const { type, payload } = action;
    switch (type) {
      case ACTION_TYPES.SET_DETAILS: {
        draft.details = payload;

        return;
      }

      default: {
        throw new Error(`Unhandled action type: ${action.type}`);
      }
    }
  }
);

export default detailsReducer;
