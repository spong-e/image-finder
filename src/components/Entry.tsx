import { FunctionComponent } from "react";
import { useFormik } from "formik";
import { useNavigate } from "react-router-dom";
import { Form, Input, Radio } from "semantic-ui-react";
import * as Yup from "yup";

import { ROUTES } from "../constants";
import { useDetailsActions } from "../providers";

const topicOptions = [
  { label: "Travel", value: "travel" },
  { label: "Cars", value: "cars" },
  { label: "Wildlife", value: "wildlife" },
  { label: "Technology", value: "technology" },
  { label: "Other", value: "other" },
];

const validationSchema = Yup.object().shape({
  firstName: Yup.string().required("First name is required"),
  lastName: Yup.string().required("Last name is required"),
  thumbnail: Yup.string(),
  topic: Yup.string(),
  otherTopic: Yup.string().when("topic", {
    is: (val: string) => val === "other",
    then: () => Yup.string().required("Other topic is required"),
  }),
});

const Entry: FunctionComponent = () => {
  const { setDetails } = useDetailsActions();
  const navigate = useNavigate();

  const formik = useFormik({
    initialValues: {
      firstName: "",
      lastName: "",
      topic: "",
      otherTopic: "",
      thumbnail: "",
    },
    validationSchema: validationSchema,
    onSubmit: (values) => {
      if (values.topic === "other") values.topic = values.otherTopic;

      const details: Details = {
        firstName: values.firstName,
        lastName: values.lastName,
        topic: values.topic,
        thumbnail: values.thumbnail,
      };

      setDetails(details);
      navigate(ROUTES.SEARCH);
    },
  });

  const otherSelected = formik.values.topic === "other";

  return (
    <div>
      <Form onSubmit={formik.handleSubmit}>
        <Form.Group widths="equal">
          <Form.Field
            id="firstName"
            control={Input}
            label="First name"
            placeholder="First name"
            value={formik.values.firstName}
            onChange={formik.handleChange}
            data-testid="firstNameInput"
            error={
              formik.touched.firstName &&
              Boolean(formik.errors.firstName) && {
                pointing: "above",
                content: formik.errors.firstName,
              }
            }
          />

          <Form.Field
            id="lastName"
            control={Input}
            label="Last name"
            placeholder="Last name"
            value={formik.values.lastName}
            onChange={formik.handleChange}
            data-testid="lastNameInput"
            error={
              formik.touched.lastName &&
              Boolean(formik.errors.lastName) && {
                pointing: "above",
                content: formik.errors.lastName,
              }
            }
          />
        </Form.Group>
        <Form.Group inline>
          {topicOptions.map((topic) => {
            return (
              <Form.Field key={`topic_${topic.value}`}>
                <Radio
                  id={`topic_${topic.value}`}
                  label={topic.label}
                  name="topic"
                  value={topic.value}
                  checked={formik.values.topic === topic.value}
                  onChange={formik.handleChange}
                  data-testid={`topic_${topic.value}Radio`}
                />
              </Form.Field>
            );
          })}
        </Form.Group>
        <Form.Group widths="equal">
          {otherSelected && (
            <Form.Field
              id="otherTopic"
              control={Input}
              label="Other topic"
              placeholder="Other topic"
              onChange={formik.handleChange}
              data-testid="otherTopicInput"
              error={
                formik.touched.otherTopic &&
                Boolean(formik.errors.otherTopic) && {
                  content: formik.errors.otherTopic,
                }
              }
            />
          )}
        </Form.Group>

        <Form.Button primary type="submit" data-testid="submitBtn">
          Search
        </Form.Button>
      </Form>
    </div>
  );
};

export default Entry;
