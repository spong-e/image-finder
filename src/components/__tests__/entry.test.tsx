import { MemoryRouter } from "react-router-dom";

import { act, fireEvent, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { wait } from "@testing-library/user-event/dist/utils";

import { ROUTES } from "../../constants";
import { useDetailsActions } from "../../providers";
import Entry from "../Entry";

import "@testing-library/jest-dom";

const details: Details = {
  firstName: "Test",
  lastName: "User",
  thumbnail: "",
  topic: "cars",
};

const mockedNavigate = jest.fn();

const mockedSetDetails = jest.fn();

const mockDetailsActions = {
  setDetails: (details: Details) => mockedSetDetails(details),
};

jest.mock("../../providers");

jest.mock("react-router-dom", () => ({
  ...(jest.requireActual("react-router-dom") as any),
  useNavigate: () => mockedNavigate,
}));

describe("Entry component", () => {
  beforeEach(async () => {
    (useDetailsActions as jest.Mock).mockReturnValue(mockDetailsActions);
  });
  describe("Component renders as expected", () => {
    it("When component loads, renders as expect", async () => {
      const { container } = render(
        <MemoryRouter>
          <Entry />
        </MemoryRouter>
      );

      expect(container).toMatchSnapshot();
    });
  });

  it("When form completed and submit clicked, setDetails function is called", async () => {
    const { container } = render(
      <MemoryRouter>
        <Entry />
      </MemoryRouter>
    );

    await act(async () => {
      const firstNameInput = await setInputValue(
        "firstNameInput",
        details.firstName
      );

      const lastNameInput = await setInputValue(
        "lastNameInput",
        details.lastName
      );

      const topicInput = await clickRadioInput("topic_carsRadio");

      expect(firstNameInput).toHaveValue(details.firstName);
      expect(lastNameInput).toHaveValue(details.lastName);
      expect(topicInput.checked).toEqual(true);
    });
    expect(container).toMatchSnapshot();

    await act(async () => {
      const submitButtonWrapper = await screen.findByTestId("submitBtn");

      fireEvent.click(submitButtonWrapper);

      await wait();

      expect(mockedSetDetails).toHaveBeenCalledWith(details);

      expect(mockedNavigate).toHaveBeenCalledWith(ROUTES.SEARCH);
    });
  });
});

async function setInputValue(testId: string, value: string) {
  const inputWrapper = await screen.findByTestId(testId);

  const input = inputWrapper.firstChild as HTMLInputElement;

  fireEvent.change(input, {
    target: { value: value },
  });

  return input;
}

async function clickRadioInput(testId: string) {
  const inputWrapper = await screen.findByTestId(testId);

  const input = inputWrapper.firstChild as HTMLInputElement;

  userEvent.click(input);

  return input;
}
