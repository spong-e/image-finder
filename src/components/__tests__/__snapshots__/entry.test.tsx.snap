// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entry component Component renders as expected When component loads, renders as expect 1`] = `
<div>
  <form>
    <div
      class="row"
    >
      <div
        class="field"
      >
        <label
          for="firstName"
        >
          First name
        </label>
        <input
          class=""
          data-testid="firstNameInput"
          id="firstName"
          name="firstName"
          placeholder="First name"
          type="text"
          value=""
        />
      </div>
      <div
        class="field"
      >
        <label
          for="lastName"
        >
          Last name
        </label>
        <input
          class=""
          data-testid="lastNameInput"
          id="lastName"
          name="lastName"
          placeholder="Last name"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="topicOptions"
    >
      <fieldset>
        <div
          class="floatBlock"
        >
          <label
            for="topic_travel"
          >
            <input
              data-testid="topic_travelRadio"
              id="topic_travel"
              name="topic"
              type="radio"
              value="travel"
            />
            Travel
          </label>
        </div>
        <div
          class="floatBlock"
        >
          <label
            for="topic_cars"
          >
            <input
              data-testid="topic_carsRadio"
              id="topic_cars"
              name="topic"
              type="radio"
              value="cars"
            />
            Cars
          </label>
        </div>
        <div
          class="floatBlock"
        >
          <label
            for="topic_wildlife"
          >
            <input
              data-testid="topic_wildlifeRadio"
              id="topic_wildlife"
              name="topic"
              type="radio"
              value="wildlife"
            />
            Wildlife
          </label>
        </div>
        <div
          class="floatBlock"
        >
          <label
            for="topic_technology"
          >
            <input
              data-testid="topic_technologyRadio"
              id="topic_technology"
              name="topic"
              type="radio"
              value="technology"
            />
            Technology
          </label>
        </div>
        <div
          class="floatBlock"
        >
          <label
            for="topic_other"
          >
            <input
              data-testid="topic_otherRadio"
              id="topic_other"
              name="topic"
              type="radio"
              value="other"
            />
            Other
          </label>
        </div>
      </fieldset>
    </div>
    <div
      class="row"
    >
      <div
        class="field"
      />
    </div>
    <div
      class="row"
    >
      <button
        class="button"
        data-testid="submitBtn"
        type="submit"
      >
        Search
      </button>
    </div>
  </form>
</div>
`;

exports[`Entry component When form 'Other' topic selected, other input appears 1`] = `
<div>
  <form>
    <div
      class="row"
    >
      <div
        class="field"
      >
        <label
          for="firstName"
        >
          First name
        </label>
        <input
          class=""
          data-testid="firstNameInput"
          id="firstName"
          name="firstName"
          placeholder="First name"
          type="text"
          value=""
        />
      </div>
      <div
        class="field"
      >
        <label
          for="lastName"
        >
          Last name
        </label>
        <input
          class=""
          data-testid="lastNameInput"
          id="lastName"
          name="lastName"
          placeholder="Last name"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="topicOptions"
    >
      <fieldset>
        <div
          class="floatBlock"
        >
          <label
            for="topic_travel"
          >
            <input
              data-testid="topic_travelRadio"
              id="topic_travel"
              name="topic"
              type="radio"
              value="travel"
            />
            Travel
          </label>
        </div>
        <div
          class="floatBlock"
        >
          <label
            for="topic_cars"
          >
            <input
              data-testid="topic_carsRadio"
              id="topic_cars"
              name="topic"
              type="radio"
              value="cars"
            />
            Cars
          </label>
        </div>
        <div
          class="floatBlock"
        >
          <label
            for="topic_wildlife"
          >
            <input
              data-testid="topic_wildlifeRadio"
              id="topic_wildlife"
              name="topic"
              type="radio"
              value="wildlife"
            />
            Wildlife
          </label>
        </div>
        <div
          class="floatBlock"
        >
          <label
            for="topic_technology"
          >
            <input
              data-testid="topic_technologyRadio"
              id="topic_technology"
              name="topic"
              type="radio"
              value="technology"
            />
            Technology
          </label>
        </div>
        <div
          class="floatBlock"
        >
          <label
            for="topic_other"
          >
            <input
              data-testid="topic_otherRadio"
              id="topic_other"
              name="topic"
              type="radio"
              value="other"
            />
            Other
          </label>
        </div>
      </fieldset>
    </div>
    <div
      class="row"
    >
      <div
        class="field"
      >
        <label
          for="otherTopic"
        >
          Other topic
        </label>
        <input
          data-testid="otherTopicInput"
          id="otherTopic"
          name="otherTopic"
          placeholder="Other topic"
        />
      </div>
    </div>
    <div
      class="row"
    >
      <button
        class="button"
        data-testid="submitBtn"
        type="submit"
      >
        Search
      </button>
    </div>
  </form>
</div>
`;

exports[`Entry component When form completed and submit clicked, setDetails function is called 1`] = `
<div>
  <form>
    <div
      class="row"
    >
      <div
        class="field"
      >
        <label
          for="firstName"
        >
          First name
        </label>
        <input
          class=""
          data-testid="firstNameInput"
          id="firstName"
          name="firstName"
          placeholder="First name"
          type="text"
          value="Test"
        />
      </div>
      <div
        class="field"
      >
        <label
          for="lastName"
        >
          Last name
        </label>
        <input
          class=""
          data-testid="lastNameInput"
          id="lastName"
          name="lastName"
          placeholder="Last name"
          type="text"
          value="User"
        />
      </div>
    </div>
    <div
      class="topicOptions"
    >
      <fieldset>
        <div
          class="floatBlock"
        >
          <label
            for="topic_travel"
          >
            <input
              data-testid="topic_travelRadio"
              id="topic_travel"
              name="topic"
              type="radio"
              value="travel"
            />
            Travel
          </label>
        </div>
        <div
          class="floatBlock"
        >
          <label
            for="topic_cars"
          >
            <input
              data-testid="topic_carsRadio"
              id="topic_cars"
              name="topic"
              type="radio"
              value="cars"
            />
            Cars
          </label>
        </div>
        <div
          class="floatBlock"
        >
          <label
            for="topic_wildlife"
          >
            <input
              data-testid="topic_wildlifeRadio"
              id="topic_wildlife"
              name="topic"
              type="radio"
              value="wildlife"
            />
            Wildlife
          </label>
        </div>
        <div
          class="floatBlock"
        >
          <label
            for="topic_technology"
          >
            <input
              data-testid="topic_technologyRadio"
              id="topic_technology"
              name="topic"
              type="radio"
              value="technology"
            />
            Technology
          </label>
        </div>
        <div
          class="floatBlock"
        >
          <label
            for="topic_other"
          >
            <input
              data-testid="topic_otherRadio"
              id="topic_other"
              name="topic"
              type="radio"
              value="other"
            />
            Other
          </label>
        </div>
      </fieldset>
    </div>
    <div
      class="row"
    >
      <div
        class="field"
      />
    </div>
    <div
      class="row"
    >
      <button
        class="button"
        data-testid="submitBtn"
        type="submit"
      >
        Search
      </button>
    </div>
  </form>
</div>
`;

exports[`Entry component When form partly completed and submit clicked, validation fires 1`] = `
<div>
  <form>
    <div
      class="row"
    >
      <div
        class="field"
      >
        <label
          for="firstName"
        >
          First name
        </label>
        <input
          class=""
          data-testid="firstNameInput"
          id="firstName"
          name="firstName"
          placeholder="First name"
          type="text"
          value=""
        />
      </div>
      <div
        class="field"
      >
        <label
          for="lastName"
        >
          Last name
        </label>
        <input
          class=""
          data-testid="lastNameInput"
          id="lastName"
          name="lastName"
          placeholder="Last name"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="topicOptions"
    >
      <fieldset>
        <div
          class="floatBlock"
        >
          <label
            for="topic_travel"
          >
            <input
              data-testid="topic_travelRadio"
              id="topic_travel"
              name="topic"
              type="radio"
              value="travel"
            />
            Travel
          </label>
        </div>
        <div
          class="floatBlock"
        >
          <label
            for="topic_cars"
          >
            <input
              data-testid="topic_carsRadio"
              id="topic_cars"
              name="topic"
              type="radio"
              value="cars"
            />
            Cars
          </label>
        </div>
        <div
          class="floatBlock"
        >
          <label
            for="topic_wildlife"
          >
            <input
              data-testid="topic_wildlifeRadio"
              id="topic_wildlife"
              name="topic"
              type="radio"
              value="wildlife"
            />
            Wildlife
          </label>
        </div>
        <div
          class="floatBlock"
        >
          <label
            for="topic_technology"
          >
            <input
              data-testid="topic_technologyRadio"
              id="topic_technology"
              name="topic"
              type="radio"
              value="technology"
            />
            Technology
          </label>
        </div>
        <div
          class="floatBlock"
        >
          <label
            for="topic_other"
          >
            <input
              data-testid="topic_otherRadio"
              id="topic_other"
              name="topic"
              type="radio"
              value="other"
            />
            Other
          </label>
        </div>
      </fieldset>
    </div>
    <div
      class="row"
    >
      <div
        class="field"
      />
    </div>
    <div
      class="row"
    >
      <button
        class="button"
        data-testid="submitBtn"
        type="submit"
      >
        Search
      </button>
    </div>
  </form>
</div>
`;
